### PR TITLE
Add "distinctHost" constraint

### PR DESCRIPTION
**Pull request by dadgar**
This PR adds a distinctHost constraint that can be applied to either Task Groups or Jobs.

### CODE REVIEW

File: website/source/docs/jobspec/index.html.md [hunk 1]
```diff
@@ -237,6 +237,14 @@ The `constraint` object supports the following keys:
   the attribute. This sets the operator to "regexp" and the `value`
   to the regular expression.
 
+* `unique` - Unique accepts a boolean value and can be used to mark a Job or
+  a Task Group as requiring placement on unique nodes. If the `unique`
+  constraint is placed on a Job, all of it's Task Groups must be placed on
```
Comment (cbednarski) [inline website/source/docs/jobspec/index.html.md]: `it's` should not have an apostrophe

File: website/source/docs/jobspec/index.html.md [hunk 2]
```diff
@@ -237,6 +237,18 @@ The `constraint` object supports the following keys:
   the attribute. This sets the operator to "regexp" and the `value`
   to the regular expression.
 
+* `distinctHosts` - `distinctHosts` accepts a boolean `true`. The default is
+  `false`.
+
+  When `distinctHosts is `true` at the Job level, each instance of all Task
```
Comment (dimfeld) [inline website/source/docs/jobspec/index.html.md]: Missing backtick on distinctHosts

File: scheduler/feasible.go [hunk 3]
```diff
@@ -257,6 +368,14 @@ func resolveConstraintTarget(target string, node *structs.Node) (interface{}, bo
 
 // checkConstraint checks if a constraint is satisfied
 func checkConstraint(ctx Context, operand string, lVal, rVal interface{}) bool {
+	// Check for constraints not handled by this iterator.
+	switch operand {
```
Comment (achanda) [inline scheduler/feasible.go]: Why not merge this with the next one?

File: scheduler/feasible.go [hunk 4]
```diff
@@ -150,6 +150,117 @@ func (iter *DriverIterator) hasDrivers(option *structs.Node) bool {
...
+}
+
+// satisfiesDistinctHosts checks if the node satisfies a distinctHosts
+// constraint either specified at the job level or the TaskGroup level.
+func (iter *DynamicConstraintIterator) satisfiesDistinctHosts(option *structs.Node, job bool) bool {
+	// Get the proposed allocations
+	proposed, err := iter.ctx.ProposedAllocs(option.ID)
+	if err != nil {
+		iter.ctx.Logger().Printf(
+			"[ERR] sched.dynamic-constraint: failed to get proposed allocations: %v", err)
```
Comment (achanda) [inline scheduler/feasible.go]: s/sched/scheduler/g

The other log statement in this file says scheduler

File: scheduler/feasible.go (hunk 3 above)
Comment (dadgar) [inline scheduler/feasible.go]: As we add more constraints that are not handled here I want it to be easily distinguishable which iterator is handling which constraints.

Comment (achanda) [inline scheduler/feasible.go]: :+1:

File: scheduler/feasible.go (hunk 4 above)
Comment (dadgar) [inline scheduler/feasible.go]: Good catch

File: jobspec/parse.go [hunk 5]
```diff
@@ -256,6 +257,20 @@ func parseConstraints(result *[]*structs.Constraint, obj *hclobj.Object) error {
 			m["RTarget"] = constraint
 		}
 
+		if value, ok := m["distinctHosts"]; ok {
```
Comment (armon) [inline jobspec/parse.go]: Nothing in HCL uses camal case

File: scheduler/feasible.go [hunk 6]
```diff
@@ -150,6 +150,117 @@ func (iter *DriverIterator) hasDrivers(option *structs.Node) bool {
...
+	iter.jobDistinctHosts = iter.hasDistinctHostsConstraint(job.Constraints)
+}
+
+func (iter *DynamicConstraintIterator) hasDistinctHostsConstraint(constraints []*structs.Constraint) bool {
+	if constraints == nil {
+		return false
+	}
+
+	for _, con := range constraints {
+		if con.Operand == "distinctHosts" {
```
Comment (armon) [inline scheduler/feasible.go]: Make this a const in `structs`

File: scheduler/feasible.go [hunk 7]
```diff
@@ -150,6 +150,117 @@ func (iter *DriverIterator) hasDrivers(option *structs.Node) bool {
...
+			return true
+		}
+	}
+	return false
+}
+
+func (iter *DynamicConstraintIterator) Next() *structs.Node {
+	if iter.job == nil {
```
Comment (armon) [inline scheduler/feasible.go]: No need to redo these checks as the entire stack is broken if no job or TG

File: scheduler/feasible.go [hunk 8]
```diff
@@ -150,6 +150,117 @@ func (iter *DriverIterator) hasDrivers(option *structs.Node) bool {
...
+			continue
+		}
+
+		return option
+	}
+}
+
+// satisfiesDistinctHosts checks if the node satisfies a distinctHosts
+// constraint either specified at the job level or the TaskGroup level.
+func (iter *DynamicConstraintIterator) satisfiesDistinctHosts(option *structs.Node, job bool) bool {
```
Comment (armon) [inline scheduler/feasible.go]: `job` argument is not necessary

File: scheduler/feasible.go [hunk 9]
```diff
@@ -150,6 +150,117 @@ func (iter *DriverIterator) hasDrivers(option *structs.Node) bool {
...
+	// Get the proposed allocations
+	proposed, err := iter.ctx.ProposedAllocs(option.ID)
```
Comment (armon) [inline scheduler/feasible.go]: Investigate caching this

File: scheduler/stack.go [hunk 10]
```diff
@@ -69,8 +70,10 @@ func NewGenericStack(batch bool, ctx Context) *GenericStack {
 	// Filter on task group constraints second
 	s.taskGroupConstraint = NewConstraintIterator(ctx, s.taskGroupDrivers, nil)
 
+	s.dynamicConstraint = NewDynamicConstraintIterator(ctx, s.taskGroupConstraint)
```
Comment (armon) [inline scheduler/stack.go]: Comment what a dynamic constraint is

File: scheduler/feasible.go [hunk 11]
```diff
@@ -150,6 +150,117 @@ func (iter *DriverIterator) hasDrivers(option *structs.Node) bool {
...
+
+	// Store whether the Job or TaskGroup has a distinctHosts constraints so
+	// they don't have to be calculated every time Next() is called.
+	tgDistinctHosts  bool
+	jobDistinctHosts bool
+}
+
+// NewDynamicConstraintIterator creates a DynamicConstraintIterator from a
+// source.
+func NewDynamicConstraintIterator(ctx Context, source FeasibleIterator) *DynamicConstraintIterator {
```
Comment (armon) [inline scheduler/feasible.go]: Rename to `ProposedAllocConstraintIterator`

File: scheduler/feasible.go [hunk 12]
```diff
@@ -150,6 +150,117 @@ func (iter *DriverIterator) hasDrivers(option *structs.Node) bool {
...
+
+	// Skip the node if the task group has already been allocated on it.
+	for _, alloc := range proposed {
+		jobCollision := alloc.JobID == iter.job.ID
+		taskCollision := alloc.TaskGroup == iter.tg.Name
+
+		// If the job has a distinctHosts constraint we only need an alloc
+		// collision on the JobID but if the constraint is on the TaskGroup then
+		// we need both a job and TaskGroup collision.
+		jobInvalid := job && jobCollision
```
Comment (armon) [inline scheduler/feasible.go]: Function should not depend on the callers behavior to implicit rule out the `!job and !tg` mode.

**Comment by armon** [comment]
LGTM other than changes

**Pull request merged**